x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/queries/SemanticQueryBuilder.java

```diff
@@ -0,0 +1,296 @@
+    }
+
+    public static SemanticQueryBuilder fromXContent(XContentParser parser) throws IOException {
+        return PARSER.apply(parser, null);
+    }
+
+    @Override
+    protected void doXContent(XContentBuilder builder, Params params) throws IOException {
+        builder.startObject(NAME);
+        builder.field(FIELD_FIELD.getPreferredName(), fieldName);
+        builder.field(QUERY_FIELD.getPreferredName(), query);
+        boostAndQueryNameToXContent(builder);
+        builder.endObject();
+    }
+
+    @Override
+    protected QueryBuilder doRewrite(QueryRewriteContext queryRewriteContext) {
+        SearchExecutionContext searchExecutionContext = queryRewriteContext.convertToSearchExecutionContext();
+        if (searchExecutionContext != null) {
+            return doRewriteBuildSemanticQuery(searchExecutionContext);
+        }
+
+        return doRewriteGetInferenceResults(queryRewriteContext);
+    }
+
+    private QueryBuilder doRewriteBuildSemanticQuery(SearchExecutionContext searchExecutionContext) {
+        if (inferenceResults == null) {
+            throw new IllegalStateException("Query builder must have inference results before rewriting to another query type");
+        }
+
+        MappedFieldType fieldType = searchExecutionContext.getFieldType(fieldName);
+        if (fieldType == null) {
+            return new MatchNoneQueryBuilder();
+        } else if (fieldType instanceof SemanticTextFieldMapper.SemanticTextFieldType semanticTextFieldType) {
+            List<? extends InferenceResults> inferenceResultsList = inferenceResults.transformToCoordinationFormat();
+            if (inferenceResultsList.isEmpty()) {
+                throw new IllegalArgumentException("No inference results retrieved for field [" + fieldName + "]");
+            } else if (inferenceResultsList.size() > 1) {
+                // TODO: How to handle multiple inference results?
+                throw new IllegalArgumentException(
+                    inferenceResultsList.size() + " inference results retrieved for field [" + fieldName + "]"
+                );
+            }
+
+            InferenceResults inferenceResults = inferenceResultsList.get(0);
+            return semanticTextFieldType.semanticQuery(inferenceResults, boost(), queryName());
+        } else {
+            throw new IllegalArgumentException(
+                "Field [" + fieldName + "] of type [" + fieldType.typeName() + "] does not support " + NAME + " queries"
+            );
+        }
+    }
+
+    private SemanticQueryBuilder doRewriteGetInferenceResults(QueryRewriteContext queryRewriteContext) {
+        if (inferenceResults != null) {
+            return this;
+        }
+
+        if (inferenceResultsSupplier != null) {
+            InferenceServiceResults inferenceResults = inferenceResultsSupplier.get();
+            return inferenceResults != null ? new SemanticQueryBuilder(this, inferenceResultsSupplier, inferenceResults) : this;
```

Comment:
```suggestion
            return inferenceResults != null ? new SemanticQueryBuilder(this, null, inferenceResults) : this;
```

Seems like the supplier should be null here.